### ISSUE TITLE
How to best implement an output body that depends on an input parameter? 

### ISSUE THREAD

**Issue by fmeriaux** [softwaremill/tapir#2244]
**Use Case:**

We have an API that has a JSON output. An input parameter allows to filter the returned JSON fields 

The only way to implement this as far as I know is to pass the formatting parameter into the server logic which will then have to encapsulate it in the return type with an appropriate codec.

We have the same problem if we want to recover the input url to build output headers (example Link prev/next headers). 

It's quite verbose, and does it really have its place in a "logic" method? 

Example:

```scala
myEndpoint
  .in(EndpointInputs.searchRequest)
  .in(EndpointInputs.jsonFilterRequest)
  .in(EndpointInputs.extractUri)
  .out(
        headers
          .and(statusCode)
          .and(jsonBody[Seq[FilteredJson[T]]])
          .map(PagedSearchResult.mapping[T])
   )
   .serverLogic { case (searchRequest, jsonFilterRequest, uri)
     myRepository.findPageable(searchRequest)
      .map {
        PagedSearchResult(
          maxSize = maxSize,
          uri = uri,
          jsonFilterRequest = jsonFilterRequest
        )
      }
   }


```

**Comment by adamw** [comment]
Yes, I'd say that the "logic" is what translates inputs to outputs :) That's exactly what the signature says: `I => F[Either[E,  O]]`.

If you have some common logic that is repeated across multiple endpoints, you may try extracting this as a function and then composing with the rest. E.g. `paging: P => O1` composed with `I => Either[E | O2]` might give you `(P, I) => Either[E, (O1, O2)]` (with a helper function, and assuming you structure the endpoints inputs/outputs this way.

**Comment by fmeriaux** [comment]
Thanks for the answer.

Indeed, it makes sense.

Nevertheless I think it could be interesting and be an added value for Tapir to embed some REST API standards.

Why not this kind of features:

- the comparison of an ETag that we have inputs like IfMatch/IfNoneMatch, like: https://doc.akka.io/docs/akka-http/current/routing-dsl/directives/cache-condition-directives/conditional.html
- self-prefixed outputs with the request url, for example to generate standard output headers like Location, Link etc. 
- ... 

The idea for Tapir would be to facilitate as much as possible the implementation of standards to build efficient REST APIs and stay focused on a purely business server logic. I think this is the core idea of this library, what do you think about it ?

**Comment by adamw** [comment]
Taking the etag/conditional headers as an example, I think we implement something similar for the static content endpoints: https://github.com/softwaremill/tapir/blob/master/core/src/main/scalajvm/sttp/tapir/static/TapirStaticContentEndpoints.scala

Akka has a different model where you don't need to capture the metadata upfront, hence some of it might be conditional. This is not possible with tapir where the structure of the endpoint is statically defined. Plus this is a strictly server-only thing, and in tapir the endpoint description is usable both for servers & clients.

I don't see how we could somehow support sth like the `conditional` directive. You could define reusable functions which translate partial input to partial output, but that's about it I'm afraid. Of course, my imagination could be failing and maybe there is something that could be done :)

**Comment by fmeriaux** [comment]
Indeed, the static part of Tapir meets a lot of REST standards. 

The problem with this implementation is that it is very restricted to a "static" use, which is consistent with its name. 

 It might be possible to evolve the current implementation to something more "resource" than "static", with a server logic signature like: 

` I => Future[Either[RessourceError, RessourceContent[T]] `

The idea of this feature request is integrate into Tapir the "standard" endpoints / parameters / outputs that are found in most common REST usages to make Tapir very easy to adopt. 

We agree that this can be implemented outside of Tapir, but it is an additional workload for a common need, which is why I think it would belong in the lib 🙂

**Comment by adamw** [comment]
Maybe this could work - if you would have some specific suggestions as to what the structure of such REST endpoints might be, and how the server logic helper function might work, please post them here :)

**Comment by fmeriaux** [comment]
I will try to make an example of how we could use a `ResourceEndpoint` with different usage and you can give me your opinion :)

**Comment by fmeriaux** [comment]
**I.** 

A first step could be to make methods implemented here public: 

https://github.com/softwaremill/tapir/blob/master/core/src/main/scalajvm/sttp/tapir/static/TapirStaticContentEndpoints.scala#L36

Why not in an object named for example `EndpointResourceIO` ?

📍 The goal would be to allow someone implementing it on his side to have standard inputs already ready.

Exemple
```scala
myCustomEndpoint.in(EndpointResourceIO.IfNoneMatch)
```

**II.** 

For the tag comparison part, to my knowledge Tapir does not expose it (?)

Either in Tapir or directly in sttp-model, the Etag class should have a `compare(tag: String): Boolean` which returns true if the tags match (taking into account the weak factor of the standard). 

📍 For this function, the objective would be to facilitate the server logic around the Etag standards in Tapir by having the necessary tools. 

III.

Finally to go further, why not class of this type:

ResourceHeadResult (variants: Found, NotModified etc)... 

With of course the necessary codecs to reference it in an `out` easily and only have to implement a logic that returns the different variants. 

📍 The objective would then be to simply assemble the various inputs/outputs available and focus on the server logic. 

It might be possible to go further in the future by making common logic available but I think it is premature at this point.

What is your opinion?
